File: packages/core/src/features/tier2.ts
import type { KeystrokeEvent } from "../types/events";
import type { Tier2Features } from "../types/features";

export function extractTier2(events: KeystrokeEvent[]): Tier2Features {
  const keydowns = events.filter(
    (e) => e.event_type === "KeyDown" && !e.repeat,
  );

  if (keydowns.length === 0) {
    return {
      backspace_rate: 0,
      delete_rate: 0,
      error_burst_count: 0,
      mean_error_burst_length: 0,
      immediate_correction_ratio: 0,
      revision_ratio: 0,
    };
  }

  const total = keydowns.length;
  const backspaceCount = keydowns.filter(
    (e) => e.key === "Backspace",
  ).length;
  const deleteCount = keydowns.filter((e) => e.key === "Delete").length;
  const correctionCount = backspaceCount + deleteCount;

  // Error bursts: consecutive sequences of backspace/delete keys
  const bursts: number[] = [];
  let currentBurst = 0;
  let immediateCorrections = 0;

  for (let i = 0; i < keydowns.length; i++) {
    const kd = keydowns[i];
    if (kd.key === "Backspace" || kd.key === "Delete") {
      currentBurst++;

      // Immediate correction: backspace right after a character key
      if (i > 0) {
        const prev = keydowns[i - 1];
        if (
          prev.key !== "Backspace" &&
          prev.key !== "Delete" &&
          prev.key.length === 1
        ) {
          const gap = kd.timestamp - prev.timestamp;
          if (gap < 500) {
            immediateCorrections++;
          }
        }
      }
    } else {
      if (currentBurst > 0) {
        bursts.push(currentBurst);
        currentBurst = 0;
      }
    }
  }
  if (currentBurst > 0) {
    bursts.push(currentBurst);
  }

  const meanBurstLen =
    bursts.length === 0
      ? 0
      : bursts.reduce((sum, b) => sum + b, 0) / bursts.length;

  return {
    backspace_rate: backspaceCount / total,
    delete_rate: deleteCount / total,
    error_burst_count: bursts.length,
    mean_error_burst_length: meanBurstLen,
    immediate_correction_ratio:
      correctionCount > 0 ? immediateCorrections / correctionCount : 0,
    revision_ratio: correctionCount / total,
  };
}

File: apps/desktop/src-tauri/src/features/tier2.rs
use serde::{Deserialize, Serialize};
use crate::capture::events::KeystrokeEvent;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Tier2Features {
    pub backspace_rate: f64,
    pub delete_rate: f64,
    pub error_burst_count: u32,
    pub mean_error_burst_length: f64,
    pub immediate_correction_ratio: f64,
    pub revision_ratio: f64,
}

pub fn extract(events: &[KeystrokeEvent]) -> Tier2Features {
    let keydowns: Vec<&KeystrokeEvent> = events
        .iter()
        .filter(|e| e.event_type == "KeyDown" && !e.repeat)
        .collect();

    if keydowns.is_empty() {
        return Tier2Features {
            backspace_rate: 0.0,
            delete_rate: 0.0,
            error_burst_count: 0,
            mean_error_burst_length: 0.0,
            immediate_correction_ratio: 0.0,
            revision_ratio: 0.0,
        };
    }

    let total = keydowns.len() as f64;
    let backspace_count = keydowns.iter().filter(|e| e.key == "Backspace").count();
    let delete_count = keydowns.iter().filter(|e| e.key == "Delete").count();
    let correction_count = backspace_count + delete_count;

    // Error bursts: consecutive sequences of backspace/delete keys
    let mut bursts: Vec<u32> = Vec::new();
    let mut current_burst: u32 = 0;
    let mut immediate_corrections: u32 = 0;

    for (i, kd) in keydowns.iter().enumerate() {
        if kd.key == "Backspace" || kd.key == "Delete" {
            current_burst += 1;

            // Immediate correction: backspace right after a character key
            if i > 0 {
                let prev = &keydowns[i - 1];
                if prev.key != "Backspace" && prev.key != "Delete" && prev.key.len() == 1 {
                    let gap = kd.timestamp - prev.timestamp;
                    if gap < 500.0 {
                        immediate_corrections += 1;
                    }
                }
            }
        } else {
            if current_burst > 0 {
                bursts.push(current_burst);
                current_burst = 0;
            }
        }
    }
    if current_burst > 0 {
        bursts.push(current_burst);
    }

    let mean_burst_len = if bursts.is_empty() {
        0.0
    } else {
        bursts.iter().sum::<u32>() as f64 / bursts.len() as f64
    };

    Tier2Features {
        backspace_rate: backspace_count as f64 / total,
        delete_rate: delete_count as f64 / total,
        error_burst_count: bursts.len() as u32,
        mean_error_burst_length: mean_burst_len,
        immediate_correction_ratio: if correction_count > 0 {
            immediate_corrections as f64 / correction_count as f64
        } else {
            0.0
        },
        revision_ratio: correction_count as f64 / total,
    }
}
